packages/experiments/src/implementation.js

```diff
@@ -38,6 +39,14 @@ const registeredExperiments = [];
 const requiredConsent =
 	'I know using unstable features means my plugin or theme will inevitably break on the next WordPress release.';
 
+/**
+ * Called by a @wordpress package wishing to opt-in to accessing or exposing
+ * private experimental APIs.
+ *
+ * @param {string} consent    The consent string.
+ * @param {string} moduleName The name of the module that is opting in.
+ * @return {{lock: Function, unlock: Function}} An object containing the lock and unlock functions.
```

Comment:
As I mentioned in the lock()/unlock() PR, there are ways to still get autocompletion in TypeScript. It's similar scenario as with getting autocompletion for `select( store )`. Here's the basic idea:

```ts
function unlock<T>( obj:any ) : T { /* ... */ }
unlock< ExperimentalComponent >( component );
```

Now, this is an additional burden on each usage and you may have to import the experimental component **type**. We can do better than that:

```ts
type Unlockable<Self, LockedType> = Self & { LockedType: LockedType };
type UnlockReturn< Arg > =
	Arg extends {LockedType: infer T}
		? T
		: Arg;

function unlock<T>( arg: T ) : UnlockReturn<T> { return {} as any; }

// Declare the experimental component with one signature
function experimentalComponent(experimentalArg: number) {
	return '2';
}

// Declare the public component with another signature
function publicComponent(publicArg: string) {
	return 1;
}

// Export the public component with experimental type signature embedded
const exportedComponent = publicComponent as Unlockable<typeof publicComponent, typeof experimentalComponent>;

// The public component is wrapped in Unlockable, but retains the original type signature and can be
// called.
const locked = exportedComponent;

// Unlocked has the experimental type signature:
const unlocked = unlock( exportedComponent );
```

[TS Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgdgGwPYGMDWBDARgiAeAZQgQDMAaKAGVTQgBMAVcCAPigF4ojSoAyKAN5Ua9JpABcw9KOZQAvgG4AsAChQkWIhoAlCMACuAJzh4oAQUMBzKG3aqAkBesQAHsAhw6AZ0HVpjZkkASzgSCEMoBjkHewB+SJjJJ2UVVRJ9OBRgIKQ4KAzkdDwGFgAKKAwrSQYoAEooSXhCtF0DY2K2IUM9IzyBOQqfDDgQBXlVVQB6SagAEQgUBEroYAALaFdIQyCAWw9gDAQoFCQdsFz9qAB3ILWoC6gvIMs4DDaINIysnLzN8N39ocAMKnc5wfalP7bPZwA4IJySOD6HZYcL1AQObptPIAcgATDiUtE1MwkCQoFCAbDgaCLrCpjMoEyoAA9WITVIqaZzBZLbpQNbQMD6HBBFDHWng2HXW6rCpwJCCiJPF5vIwfFTpTLZXJQYWilAgs504ClfUIMUIx7AbZwSzozE9YxQACMRNU6ggZL1Iothsl+wZzKZbI5QYAoi5zoZgAL1j6DRLjVLYzc7pSYXCBbIVa93hSUfQ6PRVCc4F5Y5skDH6EawZdOOaxXWTYNNM1sLg8J7vU3-cn9hQe+SM4CEC2UywUtzmaHOUGGPG+0n69Kgj4roYMGBIHQoCF2zROxAKFh9LGsRgQj5BfdtpYQodsxpc2r+cM9yhhlBUUGvwhcDoAA6UtcgrKBmnoDgKSjat3DoCd9mnRkQ3ZecuRmJoRD3VYMBveNR2pI5PUeZ483VcRQPLWMCmw6DaPQcoqxrBCA2lWpkODOdVCAA) 

`@wordpress/experiments` can be converted to TypeScript to express these types. The consumer packages would have to express the `as Unlockable` conversion with tsdoc annotations. Perhaps this would do the trick:

```
/**
  * @type {Unlockable<typeof publicComponent, typeof experimentalComponent>}
  */
const exportedComponent = publicComponent;
```

[It seems to work for simpler types](https://www.typescriptlang.org/play?filetype=js#code/PQKhFgCgAIWgBALgTwA4FNoG8ByBXAWwCN0AnAXylmCgGMB7AOwGdFoAPaAXmgBYBuKFFAQYcJGkxYUGegDMOlMTUgMWbZN2gAiAIbbBkIA)

Reply:
Hmm, does exporting the experimental type in the package's type definitions not defeat the point of `lock()` which is that these experimental features should be hidden? We want to be able to make changes to the experimental type without much concern for backwards compatibility. Maybe it's fine—just thinking out loud.

At any rate, this isn't necessary for this PR. I only needed to add a very simple return type to `lock()` so that the generated type definitions for `@wordpress/experiments` weren't completely wrong. `CustomSelectControl` doesn't have types so `unlock( CustomSelectControl )` returning `any` isn't a problem.

It will be _eventually_ be a problem when we try to use these APIs on a typed component. I recommend we (you? 😛) open a separate PR that converts `@wordpress/experiments` to TypeScript and implements one of the approaches you describe above.

Reply:
would it make sense to prefer typing these in the official types as `any`? anything that's experimental carries zero guarantees about the types

Reply:
What do you mean? Update the JSDoc here to be `return any`?

Reply:
yeah, so `unlock()` specifically returns `any`, meaning any time we `const experimentalButton = unlock( ExperimentalButton )` we get `experimentalButton: any`

point is that we can be permissive in the types so that TypeScript doesn't tell people they are in error by using the experimental thing, but we'll preserve the unstable nature of it so that we don't accidentally lock-in a specific experimental type or thing when it might change before the types are updated in projects that use them.

Reply:
Makes sense—done.

Reply:
> point is that we can be permissive in the types so that TypeScript doesn't tell people they are in error by using the experimental thing, but we'll preserve the unstable nature of it so that we don't accidentally lock-in a specific experimental type or thing when it might change before the types are updated in projects that use them.

Assuming that the Gutenberg app will almost always use the experimental version of the component by definition, wouldn't this be a huge loss for devex and code quality in this repo? I also think accurate types are exactly what help devs identify issues when the API changes.

We are working towards providing first-party types with the components package (removing our reliance on DefinitelyTyped), which we're aiming to ship in the coming weeks. So I'm concerned that the Gutenberg app will not be able to benefit from a lot of this effort.

Reply:
> Assuming that the Gutenberg app will almost always use the experimental version of the component by definition, wouldn't this be a huge loss for devex and code quality in this repo?

There's this inherent tension we've almost all been feeling with `__experimental`, which is that _intentionally_ the code is expected to change and break. I don't think the quality and devex we're looking for is there to begin with; otherwise they wouldn't be marked with the scary `__experimental`. In other words I think types on experimental code is a case of trying to have our cake and eat it too: we want stability so that consuming code doesn't break, while simultaneously we want the freedom to break things because we're not sure what the end design should look like yet.

We've had a number of cases where the inaccurate types (particularly on experimental code which has its abnormally high rate of churn) end up rejecting valid code. That's more frustrating to me at least than when TypeScript is unable to provide more assistance.

Once we move from experimental code to stable code, presumably we'll have stronger types on the stable code and get back those devex helps we wanted with the experimental code.

Reply:
I'll leave it up to the app-side devs to decide which is preferable since it impacts them the most, but I just want to make sure the decision is made based on as much context as possible.

At least for the components package:

1. Forgoing type data for _the entire component_ just for one experimental prop seems extreme.
2. Inaccurate types are due to it being reliant on DefinitelyTyped. Once we have first-party types, the types should always be up to date, and if there's some kind of oversight it can/should easily be fixed in-repo. I've been working a lot on in-repo component TS for a while now (not DT types), and I've seen no indication that types for experimental props would be any less accurate than stable ones.

Reply:
@adamziel mentioned converting `@wordpress/experiments` into TypeScript. I imagine we can revisit this discussion then, yeah?